drivers/digital-io/max22200/iio_max22200.c

```diff
@@ -0,0 +1,1221 @@
+				     const struct iio_ch_info *channel,
+				     intptr_t priv);
+
+static int max22200_iio_read_ch_mode_available(void *dev, char *buf,
+		uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv);
+
+static int max22200_iio_read_channel_freq(void *dev, char *buf, uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv);
+
+static int max22200_iio_read_channel_freq_available(void *dev, char *buf,
+		uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv);
+
+static int max22200_iio_read_enable(void *dev, char *buf, uint32_t len,
+				    const struct iio_ch_info *channel,
+				    intptr_t priv);
+
+static int max22200_iio_read_fault_mask(void *dev, char *buf, uint32_t len,
+					const struct iio_ch_info *channel,
+					intptr_t priv);
+
+static int max22200_iio_read_chopping_freq(void *dev, char *buf, uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv);
+
+static int max22200_iio_read_chopping_freq_available(void *dev, char *buf,
+		uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv);
+
+static int max22200_iio_reg_read(struct max22200_iio_desc *, uint32_t,
+				 uint32_t *);
+
+static int max22200_iio_write_scale(void *dev, char *buf, uint32_t len,
+				    const struct iio_ch_info *channel,
+				    intptr_t priv);
+
+static int max22200_iio_write_hit(void *dev, char *buf, uint32_t len,
+				  const struct iio_ch_info *channel,
+				  intptr_t priv);
+
+static int max22200_iio_write_hold(void *dev, char *buf, uint32_t len,
+				   const struct iio_ch_info *channel,
+				   intptr_t priv);
+
+static int max22200_iio_write_channel_trig(void *dev, char *buf, uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv);
+
+static int max22200_iio_write_ch_mode(void *dev, char *buf, uint32_t len,
+				      const struct iio_ch_info *channel,
+				      intptr_t priv);
+
+static int max22200_iio_write_channel_freq(void *dev, char *buf, uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv);
+
+static int max22200_iio_write_enable(void *dev, char *buf, uint32_t len,
+				     const struct iio_ch_info *channel,
+				     intptr_t priv);
+
+static int max22200_iio_write_fault_mask(void *dev, char *buf, uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv);
+
+static int max22200_iio_write_trig_state(void *dev, char *buf, uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv);
+
+static int max22200_iio_write_chopping_freq(void *dev, char *buf, uint32_t len,
+		const struct iio_ch_info *channel,
+		intptr_t priv);
+
+static int max22200_iio_reg_write(struct max22200_iio_desc *, uint32_t,
+				  uint32_t);
+
+static struct iio_attribute max22200_attrs[] = {
+	{
+		.name = "raw",
+		.show = max22200_iio_read_raw,
+	},
+	{
+		.name = "scale",
+		.show = max22200_iio_read_scale,
+		.store = max22200_iio_write_scale,
+	},
+	{
+		.name = "scale_available",
+		.show = max22200_iio_read_scale_available,
```

Comment:
You have a function for each attribute's show/store. This works but it can be written in a more compact way.

For example, with regards to these `...available` attributes, they all return a string directly, so you could have 1 single show function that handles all available attributes and differentiates them using an ID that you pass through `iio_attribute.priv`.

So instead of having 20 functions that do more or less the same think, you end up with 3/4/5, best case 1/2 for simple devices.

Reply:
Tried to group as much as possible. I couldn't add ch_mode_available with the others since it needs its own separate priv.